// league-connect/dist/authentication.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.authenticate = exports.ClientNotFoundError = exports.InvalidPlatformError = void 0;
const child_process_1 = __importDefault(require("child_process"));
const util_1 = __importDefault(require("util"));
const exec = util_1.default.promisify(child_process_1.default.exec);
/**
 * Indicates that the application does not run on an environment that the
 * League Client supports. The Client runs on windows, linux or darwin.
 */
class InvalidPlatformError extends Error {
    constructor() {
        super('process runs on platform client does not support');
    }
}
exports.InvalidPlatformError = InvalidPlatformError;
/**
 * Indicates that the league client could not be found
 */
class ClientNotFoundError extends Error {
    constructor() {
        super('league client process could not be located');
    }
}
exports.ClientNotFoundError = ClientNotFoundError;
/**
 * Locates a League Client and retrieves the credentials for the LCU API
 * from the found process
 *
 * If options.awaitConnection is false the promise will resolve into a
 * rejection if a League Client is not running
 *
 * @param options {AuthenticationOptions} Authentication options, if any
 *
 * @throws InvalidPlatformError If the environment is not running
 * windows/linux/darwin
 */
function authenticate(options) {
    return __awaiter(this, void 0, void 0, function* () {
        function tryAuthenticate() {
            return __awaiter(this, void 0, void 0, function* () {
                const portRegex = /--app-port=([0-9]+)/;
                const passwordRegex = /--remoting-auth-token=([\w-_]+)/;
                const pidRegex = /--app-pid=([0-9]+)/;
                const command = process.platform === 'win32'
                    ? 'WMIC PROCESS WHERE name=\'LeagueClientUx.exe\' GET CommandLine'
                    : 'ps x -o args | grep \'LeagueClientUx\'';
                try {
                    const { stdout } = yield exec(command);
                    const [, port] = stdout.match(portRegex);
                    const [, password] = stdout.match(passwordRegex);
                    const [, pid] = stdout.match(pidRegex);
                    return {
                        port: Number(port),
                        pid: Number(pid),
                        password,
                    };
                }
                catch (_a) {
                    throw new ClientNotFoundError();
                }
            });
        }
        // Does not run windows/linux/darwin
        if (!['win32', 'linux', 'darwin'].includes(process.platform)) {
            throw new InvalidPlatformError();
        }
        if (options === null || options === void 0 ? void 0 : options.awaitConnection) {
            // Poll until a client is found, attempting to resolve every
            // `options.pollInterval` milliseconds
            return new Promise(function self(resolve, reject) {
                tryAuthenticate().then((result) => {
                    resolve(result);
                }).catch((_) => {
                    var _a;
                    setTimeout(self, (_a = options === null || options === void 0 ? void 0 : options.pollInterval) !== null && _a !== void 0 ? _a : 2500, resolve, reject);
                });
            });
        }
        else {
            return tryAuthenticate();
        }
    });
}
exports.authenticate = authenticate;
